src/Integration.Vsix/SolutionAnalysisRequester.cs

```diff
@@ -0,0 +1,141 @@
+//-----------------------------------------------------------------------
+// <copyright file="SolutionAnalysisRequester.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Options;
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection;
+using SonarLint.VisualStudio.Integration.Vsix.Resources;
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    internal class SolutionAnalysisRequester
+    {
+        private readonly Workspace workspace;
+        private readonly IServiceProvider serviceProvider;
+
+        private object optionService;
+        private Option<bool> option;
+        private MethodInfo optionServiceSetOptionsMethod;
+        private MethodInfo optionServiceGetOptionsMethod;
+
+        public SolutionAnalysisRequester(IServiceProvider serviceProvider, Workspace workspace)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            this.serviceProvider = serviceProvider;
+            this.workspace = workspace;
+            this.FindOptionService();
+            this.FindFullSolutionAnalysisOption();
+        }
+
+        private void FindOptionService()
+        {
+            const string optionServiceTypeName = "Microsoft.CodeAnalysis.Options.IOptionService";
+            const string CodeAnalysisWorkspacesAssemblyName = "Microsoft.CodeAnalysis.Workspaces";
+
+            var optionServiceType = typeof(PerLanguageOption<int>).Assembly.GetType(optionServiceTypeName, false);
+            if (optionServiceType == null)
+            {
+                VsShellUtils.WriteToGeneralOutputPane(this.serviceProvider, Strings.MissingResourceAtLocation,
+                    optionServiceTypeName, CodeAnalysisWorkspacesAssemblyName);
+                Debug.Fail($"{optionServiceTypeName} could not be found in {CodeAnalysisWorkspacesAssemblyName}");
+                return;
+            }
+
+            this.optionService = this.workspace.Services.GetType()
+                ?.GetMethod("GetService")
```

Comment:
Worth checking if they implement IServiceProvider underneath, since you could use that instead reflection (( IServiceProvider)this.workspace.Services).GetType(optionServiceType)